Add suggestion to remove `derive()` if invoked macro is non-derive

NotStirred opened the pull request:
Adds to the existing `expected derive macro, found {}` error message:
```
help: remove the surrounding "derive()":
  --> $DIR/macro-path-prelude-fail-4.rs:1:3
   |
LL | #[derive(inline)]
   |   ^^^^^^^      ^
```

This suggestion will either fix the issue, in the case that the macro was valid, or provide a better error message if not

Not ready for merge yet, as the highlighted span is only valid for trivial formatting. Is there a nice way to get the parent span of the macro path within `smart_resolve_macro_path`?

Closes #109589

NotStirred committed:
Add suggestion to remove derive() if invoked macro is non-derive

workingjubilee commented:
> My trouble comes with knowing what the macro is invoked on, (function, struct), and sending the help message selectively based on that. Is there a way to get this information in smart_resolve_macro_path, or is my approach wrong and the error should be elsewhere?

Hm. What do you mean by "sending the help message selectively", exactly?

Though, if this was known to be 100% correct, which I imagine selectivity might help, it would be nice if this had `Applicability::MachineApplicable` in those cases.

NotStirred commented:
Sorry for the confusion, originally I'd thought the help message was invalid for some cases, but it looks to be fine.

Based on tests all cases are valid. Though the `Span` adjustment code is definitely wrong as I mentioned in the original PR message

cjgillot commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -544,11 +544,23 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            if kind == MacroKind::Derive && ext.macro_kind() != MacroKind::Derive {
+                // Suggest removing the derive() as the macro isn't Derive
+                let opening_span =
+                    path.span.shrink_to_lo().with_lo(path.span.lo() - rustc_span::BytePos(7));
```
This may not work when:
- there is space between the derive and the path? `#[ derive( Path ) ]`
- there are multiple paths in the same derive? `#[derive(A, B)]`
- the path comes from a macro expansion? `#[derive($path)]` in a macro with `$path` interpolation variable.

NotStirred replied:
Yes, this was part of my issues, what is the proper way to get the parent `Span`?

davidtwco replied:
I wonder if it would be better to:

- Skip this if the `Span` is from a macro, because it's probably difficult to make an appropriate suggestion in that context.
- Suggest removing the `Path` from `#[derive(Path)]` and suggest adding a new line with `#[Path]` instead?
  - This would be better for the `#[derive(A, B)]` case but worse for the `#[derive(A)]` case.

NotStirred replied:
Yes I think that's reasonable - after a lot of poking around I think it's unreasonable to get the parent span as it would require some significant changes (far more than I am comfortable with currently)

NotStirred replied:
Could you clarify what you mean by:
> Skip this if the Span is from a macro, because it's probably difficult to make an appropriate suggestion in that context.

Do you mean as span such as this? As far as I know it isn't valid
```
#[derive(foo!())]
         ^^^^^^
```

davidtwco replied:
I just mean using [`Span::from_expansion`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.from_expansion) to check whether this code was generated by a macro, since there's not much we can do with `#[derive($foo)]`.

NotStirred replied:
Oh I see - many thanks

NotStirred committed:
Update non-derive macro error message to match suggestion

It's now split between two errors, one to remove the invalid derive macro
and one suggesting adding a new non-derive macro

NotStirred committed:
Fix typo

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4568906278/jobs/8064532615) [(plain)](https://github.com/rust-lang/rust/commit/ed84d1d9052912d4d948313be72a6dd50c2baa26/checks/12408924295/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8f4b7f84864484a7bf31766abe9204da3cbe65b3)
Download action repository 'rust-lang/simpleinfra@master' (SHA:3040325909b538d8ad81ad89a04b7a91b109c313)
Complete job name: PR (x86_64-gnu-llvm-14, false, ubuntu-20.04-16core-64gb)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-14
---
-    |          not a derive macro
-    |          help: Remove from the surrounding `derive()`
+    |          ^^^^^^ not a derive macro
9    |
+ help: Remove from the surrounding `derive()`
+    |
+    |
+ LL | #[derive(inline)]
10    = help: Add as non-Derive macro
10    = help: Add as non-Derive macro
11            `#[inline]`


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/macro-path-prelude-fail-4/macro-path-prelude-fail-4.stderr
To only update this specific test, also pass `--test-args macros/macro-path-prelude-fail-4.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/macro-path-prelude-fail-4.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/macro-path-prelude-fail-4" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/macro-path-prelude-fail-4/auxiliary"
stdout: none
--- stderr -------------------------------
error: expected derive macro, found built-in attribute `inline`
  --> fake-test-src-base/macros/macro-path-prelude-fail-4.rs:1:10
   |
LL | #[derive(inline)] //~ ERROR expected derive macro, found built-in attribute `inline`
   |          ^^^^^^ not a derive macro
help: Remove from the surrounding `derive()`
  --> fake-test-src-base/macros/macro-path-prelude-fail-4.rs:1:10
   |
   |
LL | #[derive(inline)] //~ ERROR expected derive macro, found built-in attribute `inline`
   = help: Add as non-Derive macro
   = help: Add as non-Derive macro
           `#[inline]`
error: aborting due to previous error
------------------------------------------



---- [ui] tests/ui/proc-macro/macro-namespace-reserved-2.rs stdout ----
diff of stderr:

56   --> $DIR/macro-namespace-reserved-2.rs:53:10
57    |
58 LL | #[derive(my_macro_attr)]
-    |          |
-    |          not a derive macro
-    |          help: Remove from the surrounding `derive()`
+    |          ^^^^^^^^^^^^^ not a derive macro
+    |          ^^^^^^^^^^^^^ not a derive macro
63    |
+ help: Remove from the surrounding `derive()`
+    |
+    |
+ LL | #[derive(my_macro_attr)]
64    = help: Add as non-Derive macro
64    = help: Add as non-Derive macro
65            `#[my_macro_attr]`

92   --> $DIR/macro-namespace-reserved-2.rs:50:10
93    |
93    |
94 LL | #[derive(crate::my_macro)]
-    |          |
-    |          not a derive macro
-    |          help: Remove from the surrounding `derive()`
+    |          ^^^^^^^^^^^^^^^ not a derive macro
+    |          ^^^^^^^^^^^^^^^ not a derive macro
99    |
+ help: Remove from the surrounding `derive()`
+    |
+    |
+ LL | #[derive(crate::my_macro)]
100    = help: Add as non-Derive macro
100    = help: Add as non-Derive macro
101            `#[crate::my_macro]`


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/macro-namespace-reserved-2/macro-namespace-reserved-2.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/macro-namespace-reserved-2/macro-namespace-reserved-2.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args proc-macro/macro-namespace-reserved-2.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/macro-namespace-reserved-2.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/macro-namespace-reserved-2" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/macro-namespace-reserved-2/auxiliary"
stdout: none
--- stderr -------------------------------
error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:25:5
   |
LL |     my_macro!(); //~ ERROR can't use a procedural macro from the same crate that defines it

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:29:5
   |
   |
LL |     crate::my_macro_attr!(); //~ ERROR can't use a procedural macro from the same crate that defines


error: expected macro, found attribute macro `crate::my_macro_attr`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:29:5
   |
LL |     crate::my_macro_attr!(); //~ ERROR can't use a procedural macro from the same crate that defines

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:34:5
   |
   |
LL |     crate::MyTrait!(); //~ ERROR can't use a procedural macro from the same crate that defines it


error: expected macro, found derive macro `crate::MyTrait`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:34:5
   |
LL |     crate::MyTrait!(); //~ ERROR can't use a procedural macro from the same crate that defines it

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:42:3
   |
   |
LL | #[my_macro_attr] //~ ERROR can't use a procedural macro from the same crate that defines it

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:44:3
   |
   |
LL | #[MyTrait] //~ ERROR can't use a procedural macro from the same crate that defines it


error: expected attribute, found derive macro `MyTrait`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:44:3
   |
LL | #[MyTrait] //~ ERROR can't use a procedural macro from the same crate that defines it

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:53:10
   |
   |
LL | #[derive(my_macro_attr)] //~ ERROR can't use a procedural macro from the same crate that defines it


error: expected derive macro, found attribute macro `my_macro_attr`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:53:10
   |
LL | #[derive(my_macro_attr)] //~ ERROR can't use a procedural macro from the same crate that defines it
   |          ^^^^^^^^^^^^^ not a derive macro
help: Remove from the surrounding `derive()`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:53:10
   |
   |
LL | #[derive(my_macro_attr)] //~ ERROR can't use a procedural macro from the same crate that defines it
   = help: Add as non-Derive macro
   = help: Add as non-Derive macro
           `#[my_macro_attr]`
error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:56:10
   |
   |
LL | #[derive(MyTrait)] //~ ERROR can't use a procedural macro from the same crate that defines it

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:39:3
   |
   |
LL | #[crate::my_macro] //~ ERROR can't use a procedural macro from the same crate that defines it


error: expected attribute, found macro `crate::my_macro`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:39:3
   |
LL | #[crate::my_macro] //~ ERROR can't use a procedural macro from the same crate that defines it

error: can't use a procedural macro from the same crate that defines it
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:50:10
   |
   |
LL | #[derive(crate::my_macro)] //~ ERROR can't use a procedural macro from the same crate that defines


error: expected derive macro, found macro `crate::my_macro`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:50:10
   |
LL | #[derive(crate::my_macro)] //~ ERROR can't use a procedural macro from the same crate that defines
   |          ^^^^^^^^^^^^^^^ not a derive macro
help: Remove from the surrounding `derive()`
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:50:10
   |
   |
LL | #[derive(crate::my_macro)] //~ ERROR can't use a procedural macro from the same crate that defines
   = help: Add as non-Derive macro
   = help: Add as non-Derive macro
           `#[crate::my_macro]`
error: cannot find macro `my_macro_attr` in this scope
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:28:5
   |
   |
LL |     my_macro_attr!(); //~ ERROR cannot find macro `my_macro_attr` in this scope
   |
   |
   = note: `my_macro_attr` is in scope, but it is an attribute: `#[my_macro_attr]`
error: cannot find macro `MyTrait` in this scope
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:33:5
   |
   |
LL |     MyTrait!(); //~ ERROR cannot find macro `MyTrait` in this scope
   |
   |
   = note: `MyTrait` is in scope, but it is a derive macro: `#[derive(MyTrait)]`

error: cannot find attribute `my_macro` in this scope
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:38:3
   |
LL | #[my_macro] //~ ERROR cannot find attribute `my_macro` in this scope
   |
   |
   = note: `my_macro` is in scope, but it is a function-like macro

error: cannot find derive macro `my_macro` in this scope
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:48:10
   |
LL | #[derive(my_macro)] //~ ERROR cannot find derive macro `my_macro` in this scope
   |
   |
   = note: `my_macro` is in scope, but it is a function-like macro

error: cannot find derive macro `my_macro` in this scope
  --> fake-test-src-base/proc-macro/macro-namespace-reserved-2.rs:48:10
   |
LL | #[derive(my_macro)] //~ ERROR cannot find derive macro `my_macro` in this scope
   |
   |
   = note: `my_macro` is in scope, but it is a function-like macro
error: aborting due to 20 previous errors
------------------------------------------



---- [ui] tests/ui/tool-attributes/tool-attributes-misplaced-2.rs stdout ----
diff of stderr:

2   --> $DIR/tool-attributes-misplaced-2.rs:1:10
3    |
4 LL | #[derive(rustfmt::skip)]
-    |          |
-    |          not a derive macro
-    |          help: Remove from the surrounding `derive()`
+    |          ^^^^^^^^^^^^^ not a derive macro
+    |          ^^^^^^^^^^^^^ not a derive macro
9    |
+ help: Remove from the surrounding `derive()`
+    |
+    |
+ LL | #[derive(rustfmt::skip)]
10    = help: Add as non-Derive macro
10    = help: Add as non-Derive macro
11            `#[rustfmt::skip]`


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool-attributes/tool-attributes-misplaced-2/tool-attributes-misplaced-2.stderr
To only update this specific test, also pass `--test-args tool-attributes/tool-attributes-misplaced-2.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/tool-attributes/tool-attributes-misplaced-2.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool-attributes/tool-attributes-misplaced-2" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool-attributes/tool-attributes-misplaced-2/auxiliary"
stdout: none
--- stderr -------------------------------
error: expected derive macro, found tool attribute `rustfmt::skip`
  --> fake-test-src-base/tool-attributes/tool-attributes-misplaced-2.rs:1:10
   |
LL | #[derive(rustfmt::skip)] //~ ERROR expected derive macro, found tool attribute `rustfmt::skip`
   |          ^^^^^^^^^^^^^ not a derive macro
help: Remove from the surrounding `derive()`
  --> fake-test-src-base/tool-attributes/tool-attributes-misplaced-2.rs:1:10
   |
   |
LL | #[derive(rustfmt::skip)] //~ ERROR expected derive macro, found tool attribute `rustfmt::skip`
   = help: Add as non-Derive macro
   = help: Add as non-Derive macro
           `#[rustfmt::skip]`

error: expected macro, found tool attribute `rustfmt::skip`
  --> fake-test-src-base/tool-attributes/tool-attributes-misplaced-2.rs:5:5
   |
LL |     rustfmt::skip!(); //~ ERROR expected macro, found tool attribute `rustfmt::skip`

error: aborting due to 2 previous errors
------------------------------------------

```

</details>

NotStirred committed:
Bless

NotStirred committed:
Skip help messages if macro span is from macro

davidtwco commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -544,11 +544,19 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+
+            if !path.span.from_expansion() {
+                // Suggest moving the macro out of the derive() as the macro isn't Derive
+                if kind == MacroKind::Derive && ext.macro_kind() != MacroKind::Derive {
```
nit: collapse these two conditions into one

e.g. `if !path.span.from_expansion() && kind == MacroKind::Derive /* and so on... */`

davidtwco commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -544,11 +544,19 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            if !path.span.from_expansion() {
+                // Suggest moving the macro out of the derive() as the macro isn't Derive
+                if kind == MacroKind::Derive && ext.macro_kind() != MacroKind::Derive {
+                    err.span_help(path.span, "Remove from the surrounding `derive()`");
```
```suggestion
                    err.span_help(path.span, "remove from the surrounding `derive()`");
```
diagnostic messages always start with lowercase

davidtwco commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -544,11 +544,19 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
-                .struct_span_err(path.span, &msg)
-                .span_label(path.span, format!("not {} {}", article, expected))
-                .emit();
+            let mut err = self.tcx.sess.struct_span_err(path.span, &msg);
```
Could you change this to a translatable diagnostic? See #100717

davidtwco commented on tests/ui/macros/macro-path-prelude-fail-4.stderr:
```diff
@@ -3,6 +3,14 @@ error: expected derive macro, found built-in attribute `inline`
 LL | #[derive(inline)]
    |          ^^^^^^ not a derive macro
+   |
+help: Remove from the surrounding `derive()`
```
Could you add a new test case that has some of the trickier cases for this change? e.g. two derives, macro (to confirm it doesn't trigger), etc.

NotStirred replied:
As far as I can tell, this isn't valid. Was there something else you meant with macro?
```rust
macro_rules! foo {
    () => {
        Debug
    };
}

#[derive(foo!())]
struct V;
```

davidtwco replied:
```rust
macro_rules! foo {
    ($foo:expr) => {
		#[derive($foo)]
		struct V;
    };
}
```

Something more like this?

davidtwco reviewed (changes requested):
apologies for the delay in getting back to this, left some more comments

NotStirred committed:
Collapse if statements, change message to lowercase

NotStirred committed:
Add more complex test cases for new suggestion

NotStirred committed:
Rewrite added diagnostics as translatable

Start messages with lowercase

NotStirred committed:
Change type and field name to be clearer

NotStirred commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -543,12 +544,30 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            }
+
+            let mut err = self.tcx.sess.create_err(err);
+            err.span_label(path.span, format!("not {} {}", article, expected));
```
I've not added this to the translatable diagnostic as dealing with the translation (or not, depending on the language) of articles is probably out of scope for this PR

davidtwco reviewed (approved):
I'll approve this since it's been a little while and it's in a good state. Thanks for working on this.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 668a62984ac51a09987c5b1a7c1589814f7a95d4 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 668a62984ac51a09987c5b1a7c1589814f7a95d4 -->
<!-- homu: {"type":"Approved","sha":"668a62984ac51a09987c5b1a7c1589814f7a95d4","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 668a62984ac51a09987c5b1a7c1589814f7a95d4 with merge 194a0bb5d6fbdd36e0df58b4698311bb8fed6bca...
<!-- homu: {"type":"BuildStarted","head_sha":"668a62984ac51a09987c5b1a7c1589814f7a95d4","merge_sha":"194a0bb5d6fbdd36e0df58b4698311bb8fed6bca"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4661570908/jobs/8252704095)
Approved by: davidtwco
Pushing 194a0bb5d6fbdd36e0df58b4698311bb8fed6bca to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4661570908/jobs/8252704095"},"merge_sha":"194a0bb5d6fbdd36e0df58b4698311bb8fed6bca"} -->

rust-timer commented:
Finished benchmarking commit (194a0bb5d6fbdd36e0df58b4698311bb8fed6bca): [comparison URL](https://perf.rust-lang.org/compare.html?start=88fb1b922b047981fc0cfc62aa1418b4361ae72e&end=194a0bb5d6fbdd36e0df58b4698311bb8fed6bca&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=88fb1b922b047981fc0cfc62aa1418b4361ae72e&end=194a0bb5d6fbdd36e0df58b4698311bb8fed6bca&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=88fb1b922b047981fc0cfc62aa1418b4361ae72e&end=194a0bb5d6fbdd36e0df58b4698311bb8fed6bca&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 4.1%  | [4.1%, 4.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.0%, -1.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-1.0%, 0.4%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=88fb1b922b047981fc0cfc62aa1418b4361ae72e&end=194a0bb5d6fbdd36e0df58b4698311bb8fed6bca&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.8%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -5.1% | [-6.3%, -3.9%] | 2     |
| All ❌✅ (primary)                 | -0.7% | [-0.8%, -0.5%] | 2     |
</details>
